streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -1092,13 +1102,55 @@ public boolean commitRequested() {
         return commitRequested;
     }
 
+    static String encodeTimestampAndProcessorMetadata(final TopicPartition partition,
+        final long partitionTime, final ProcessorMetadata metadata) {
+        final byte[] serializedMeta = metadata.serialize(partition);
+        // Format: MAGIC_BYTE(1) + PartitionTime(8) + processMeta
+        final ByteBuffer buffer = ByteBuffer.allocate(1 + 8 + serializedMeta.length);
+        buffer.put(LATEST_MAGIC_BYTE);
+        buffer.putLong(partitionTime);
+        buffer.put(serializedMeta);
+        return Base64.getEncoder().encodeToString(buffer.array());
+    }
+
     static String encodeTimestamp(final long partitionTime) {
         final ByteBuffer buffer = ByteBuffer.allocate(9);
         buffer.put(LATEST_MAGIC_BYTE);
         buffer.putLong(partitionTime);
         return Base64.getEncoder().encodeToString(buffer.array());
     }
 
+    KeyValue<Long, ProcessorMetadata> decodeTimestampAndMeta(final String encryptedString, final TopicPartition partition) {
+        long timestamp = RecordQueue.UNKNOWN;
+        ProcessorMetadata metadata = null;
+
+        if (encryptedString.isEmpty()) {
+            return KeyValue.pair(RecordQueue.UNKNOWN, null);
+        }
+        try {
+            final ByteBuffer buffer = ByteBuffer.wrap(Base64.getDecoder().decode(encryptedString));
+            final byte version = buffer.get();
+            switch (version) {
+                case (byte) 1:
+                    timestamp = buffer.getLong();
+                    break;
+                case LATEST_MAGIC_BYTE:
+                    timestamp = buffer.getLong();
+                    final byte[] metaBytes = new byte[buffer.remaining()];
+                    buffer.get(metaBytes);
+                    metadata = ProcessorMetadata.deserialize(metaBytes, partition);
+                    break;
+                default:
+                    log.warn("Unsupported offset metadata version found. Supported version <= {}. Found version {}.",
+                            LATEST_MAGIC_BYTE, version);
+            }
+        } catch (final Exception exception) {
+            log.warn("Unsupported offset metadata found");
+        }
+
+        return KeyValue.pair(timestamp, metadata);
+    }
+
     long decodeTimestamp(final String encryptedString) {
```

Comment:
Seems this will be removed as it's replace by `decodeTimestampAndMeta` that can handle old and new format?

Reply:
Right.